_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,791 @@
+        for i in 1..24 {
+            let created_at = Utc::now().format("%Y/%m/%d %H:00:00.000 %z").to_string();
+            let created_at =
+                DateTime::parse_from_str(&created_at, "%Y/%m/%d %H:%M:%S%.3f %z").unwrap() -
+                chrono::Duration::seconds((3600 * i).into());
+
+            if !data.iter().any(|x| x.time == created_at) {
+                data.push(WebsiteStats {
+                    time: created_at.into(),
+                    uptime_pct: None,
+                });
+            }
+        }
+        data.sort_by(|a, b| b.time.cmp(&a.time));
+    }
+
+    logs.push(WebsiteInfo {
+        url: website.url,
+        alias: website.alias,
+        data,
+    });
+}
+```
+
+However, this is quite a large block of code to put in a handler function. What about if we want to be able to grab data over a longer period? Say, a month?
+
+We can do this by declaring an enum:
+
+```rust
+enum SplitBy {
+    Hour,
+    Day
+}
+```
+
+This enum will allow us to differentiate what period we want data over. We can abstract this block into its own function:
+
+```rust
+async fn get_website_uptime_stats(split_by: SplitBy, alias: &str, db: &PgPool
+) -> Vec<WebsiteStats> {
+        let mut data = match split_by {
+           SplitBy::Hour => sqlx::query_as::<_, WebsiteStats>(
+            r#"
+            SELECT date_trunc('hour', created_at) as time,
+            CAST(COUNT(case when status = 200 then 1 end) * 100 / COUNT(*) AS int2) as uptime_pct
+            FROM logs WHERE website_alias = $1
+            group by time
+            order by time asc
+            limit 24
+            "#,
+        )
+        .bind(alias)
+        .fetch_all(db)
+        .await
+        .unwrap(),
+            SplitBy::Day => sqlx::query_as::<_, WebsiteStats>(
+            r#"
+            SELECT date_trunc('day', created_at) as time,
+            CAST(COUNT(case when status = 200 then 1 end) * 100 / COUNT(*) AS int2) as uptime_pct
+            FROM logs WHERE website_alias = $1
+            group by time
+            order by time asc
+            limit 30
+            "#,
+        )
+        .bind(alias)
+        .fetch_all(db)
+        .await
+        .unwrap()
+        };
+
+    // .. rest of the function code
+}
+```
+
+The above SQL queries split data into either 24 hourly blocks or 30 day blocks, depending on the `SplitBy` variant.
+
+Then here, we can match `split_by` to get the values that we need - the number of data points we need, the date format and the number of seconds we need to wait in between
+
+```rust
+let number_of_splits = match split_by {
+            SplitBy::Hour => 24,
+            SplitBy::Day => 30
+        };
+
+        let date_format = match split_by {
+            SplitBy::Hour => "%Y/%m/%d %H:00:00.000 %z",
+            SplitBy::Day => "%Y/%m/%d 00:00:00.000 %z"
+        };
+
+        let number_of_seconds = match split_by {
+            SplitBy::Hour => 3600,
+            SplitBy::Day => 86400,
+        };
```

Comment:
These multiple matches indicate the need to rather have two functions for each enum variant